### ISSUE TITLE
Mac Build Errors

### ISSUE THREAD

**Issue by PhilipTobias** [dauxio/daux.io#211]
Getting the following errors trying to get the latest version of Daux running on Mac OS 10.13.6. I am trying to build the default docs before making any custom changes. Since I am able to build the prior Daux version in Windows 10, and used to build old versions of Daux on this Mac, I wonder if I have a Mac configuration error? Help, please.

Steps to reproduce:
1. Download and unzip the latest build of Daux
2. Run the command: daux generate
3. Errors appear in command window; the asset copying to the Static folder stops with themes folder

Error excerpt: [The initial Deprecated error appears multiple times before the following excerpt]

Deprecated: array_key_exists(): Using array_key_exists() on objects is deprecated. Use isset() or property_exists() instead in /Users/philiptobias/.composer/vendor/daux/daux.io/libs/Config.php on line 75
PHP Fatal error:  Uncaught Error: Call to undefined method Todaymade\Daux\Config::getBasePage() in /Users/philiptobias/Documents/Tests/Daux/templates/partials/navbar_content.php:6
Stack trace:
#0 /Users/philiptobias/.composer/vendor/league/plates/src/Template/Template.php(153): include()
#1 /Users/philiptobias/.composer/vendor/league/plates/src/Engine.php(277): League\Plates\Template\Template->render(NULL)
#2 /Users/philiptobias/.composer/vendor/league/plates/src/Template/Template.php(289): League\Plates\Engine->render('theme::partials...', Array)
#3 /Users/philiptobias/Documents/Tests/Daux/templates/home.php(7): League\Plates\Template\Template->insert('theme::partials...')
#4 /Users/philiptobias/.composer/vendor/league/plates/src/Template/Template.php(153): include('/Users/philipto...')
#5 /Users/philiptobias/.composer/vendor/league/plates/src/Engine.php(277): League\Plates\Template\Template->render(NULL)
#6 /Users/philiptobias/.composer/vendor/daux/daux.io/libs/Format/HTML/Template.php(75): League\Plates\Engine->render('them in /Users/philiptobias/Documents/Tests/Daux/templates/partials/navbar_content.php on line 6

**Comment by onigoetz** [comment]
Hello,

I'm not sure I understand your issue correctly
- by the looks of the deprecation error (array_key_exists in `libs/Config.php` on line 75) there is no array_key_exists in the latest version (0.14.0) at line 75 in this file. The last version containing this function at this line was 0.11.1
- You say you installed through a zip, but the path in the exception `/Users/philiptobias/.composer/vendor/daux/daux.io/` seems to point to a global install.

Thus it seems you have both a local and global install of daux.
I'd recommend upgrading the global version with `composer global update` or `composer global require daux/daux.io` 

Can you try this ?
If you need a separate version of daux for a specific project, you can install it with `composer require daux/daux.io` no need to download the zip.

**Comment by PhilipTobias** [comment]
Thanks for the prompt response. I have had various separate Daux projects, dating back to the early justinwalsh days.

The other day I downloaded the current Daux fileset zip and tried building its default docs directory. That is what is erroring out.

As you suggested, I just now tried running the command: 
composer global update

After running that command, I changed the directory back to the new project's location, then ran the command:
daux generate

That once again resulted in the errors shown in my original message above.

So how can I clean this up, so I can build your current fileset? I no longer need to support my older Daux projects, but I would like to get things working so the new Daux fileset builds. Once I can get that building without error, I can replace your docs directory with my own.

**Comment by onigoetz** [comment]
Nice to have a regular user :) 

The way the tool works now is that you only need to have the `docs` directory with markdown files and maybe a `docs/config.json` all the other files (libs, themes ...) are not needed to run Daux locally.

The installation can be done in three ways
1. The global install, upside; you do it only once, downside; you need to make sure it works for all your project (command to run is `daux generate`)
2. Locally to the project, using `composer require` will install it as a dependency to your current project. upside; each project can have a different version; downside, you might get dependencies conflicts (command to run is `vendor/bin/daux generate`)
3. Use the Docker image. upside; no dependencies to install other than Docker, leaves your environment clean. Downside; complex if you're not used to Docker.

In your case, it seems that Option 2 would be cleaner since you have multiple different projects, if it is already a PHP project, juste run `composer require daux/daux.io` otherwise create a `composer.json` containing only `{}` and run the same command.

As for the version you've downloaded, you can delete everything but the `docs` folder and it should work better already

**Comment by PhilipTobias** [comment]
Thanks, I'll give this a try when I have a moment. Option 2 does seem most practical in my situation.

I am indeed a "regular user." I delivered my first Daux documentation project using the original Justin Walsh framework. The client has since abstracted and integrated that older Daux project into their overall build process. When they build their software, it also builds the documentation - and even uses "includes" to import current code samples during the build.

My other delivered project now uses your more current framework, though not the latest. In both cases, I heavily customized the template and CSS to match their branding requirements. For instance, beyond changing the colors, I changed the search field location, added a tabbed interface for multiple code examples (C#, Java, Python), and so on. I wish I could share those as samples, but both projects are proprietary.

Anyway, I'll give your suggestions a try, and may be back with more questions. In the meantime, thank you for maintaining this wonderful documentation generator!

...pt




> On Jun 12, 2020, at 1:26 AM, Stéphane Goetz <notifications@github.com> wrote:
> 
> 
> Nice to have a regular user :)
> 
> The way the tool works now is that you only need to have the docs directory with markdown files and maybe a docs/config.json all the other files (libs, themes ...) are not needed to run Daux locally.
> 
> The installation can be done in three ways
> 
> 	• The global install, upside; you do it only once, downside; you need to make sure it works for all your project (command to run is daux generate)
> 	• Locally to the project, using composer require will install it as a dependency to your current project. upside; each project can have a different version; downside, you might get dependencies conflicts (command to run is vendor/bin/daux generate)
> 	• Use the Docker image. upside; no dependencies to install other than Docker, leaves your environment clean. Downside; complex if you're not used to Docker.
> In your case, it seems that Option 2 would be cleaner since you have multiple different projects, if it is already a PHP project, juste run composer require daux/daux.io otherwise create a composer.json containing only {} and run the same command.
> 
> As for the version you've downloaded, you can delete everything but the docs folder and it should work better already
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> 

-- 
Business/Communications start at http://www.PhilipTobias.com.
Grow your business using my technical and marketing communications -
Effective writing, graphic design, multimedia, photography, and websites.

**Comment by PhilipTobias** [comment]
Hi, sorry for the long delay. After delivering the client's last complete Daux.io project, I got moved to other projects. So I did not have a chance until now to test your suggestions about configuring any future Daux projects.

Today though I tested your Option 2, but am getting a new error.

Here's what I did:
- Created a new test folder originally containing only a docs folder.
- In a terminal window, changed directory to the new test folder.
- From that directory, ran the command: composer require daux/daux.io
- After the composer packages were installed, from the same test directory I ran the command: vendor/bin/daux generate
- Copying Static assets seems to complete OK, but the process errors out when generating. [The static folder gets created with a themes subfolder, but no other files get generated.]
- This terminal error displays:
	The error is In Config.php line 150:
	Passed variable is not an array or object

That error appears whether using the default composer.json file generated during the installation, or after editing composer.json to contain only: {}

In case I have a chance to get back to this, do you have any suggestions about getting past this new error so I can build the sample project?

Thanks in advance.   ...pt

**Comment by onigoetz** [comment]
Thanks for trying it out,

* Does that test `docs` folder contain at least two markdown files ?  for example `index.md` and `01_first_page.md` ?
* I don't get why this `composer.json` is empty ? you removed the `daux/daux.io` dependency after installing ? anyway, that shouldn't have an impact as long as it's in vendor
* Can you give me your version of php ? (`php --version`)
* I guess the version of daux that got installed is the latest ? (0.14.0)

The error seems to point to https://github.com/dauxio/daux.io/blob/d5565362c8b8edaa737337c6a5b388fb6d542587/libs/Config.php#L148-L151 so something specific to a theme. Do you have a custom theme ?

**Comment by PhilipTobias** [comment]
Here are my answers to today's questions...

• Yes, my test docs folder contains several markdown files. I simply copied and pasted the usual sample docs folder into this new once-empty project.
• The composer.json file that got installed after running your latest CLI commands was initially populated. However, since I was getting the new error with that populated file, I tried your other suggestion to remove all from that file except the {}, but that too resulted in the same new error.
• My PHP version is 7.4.6.
• I'm not sure where or how to determine the version of daux.
• This new test project should not have a custom theme. The config.json file in the test docs folder shows: "theme": "daux-blue", "template": "default". However, some of the earlier daux projects I delivered to clients did have custom themes. As best I recall, I have created custom themes using two different daux frameworks. Those projects used to build and work fine at the time, and are still on my computer. Is this new project confused by those other projects?

Let me know if you have any new suggestions. I'd love to figure this out, so I can build future projects for this client or others. Unfortunately, I don't really understand what is going on in the background with composer, dependencies, the overall build process, and so on. But it sure seems overly difficult to configure or manage several different daux projects. Thanks for all your help!

**Comment by PhilipTobias** [comment]
We may as well close this ticket. My customer moved me to other projects, so I have not had to further pursue a fix for this particular problem build on my Mac. In fact I have since sold and replaced that Mac with another. 

Today though I set up a new Daux test environment on my current Windows computer. I was able to build the new Daux project with your current file set. So if my customer needs future updates, I should be able to revise and build on my Windows computer. Or set up my new Mac if needed.

Thanks for the help along the way.   ...pt
